### PR TITLE
Add timeouts to resolvers

### PR DESCRIPTION
**Pull request by alloy**
Closes https://artsyproduct.atlassian.net/browse/PLATFORM-425

After going back and forth a few times on how to approach this (e.g. have a fixed overall timeout, but that’s weird because some queries legitimately take longer than others) I finally settled on the simplest solution, which is to have a timeout per resolver.

I’ve hardcoded a default of 5 seconds, but am open to changing that if people have opinions.

If resolvers need to deviate and take longer, they can technically do so by using a directive, except that will require generating a bit of AST because we don’t use the SDL to define out schema (dear future reader, ping me if you are in need of this).

⚠️ We should only deploy this to production while keeping a close eye out for these errors, otherwise we risk timing out queries that succeed today.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -96,12 +97,14 @@
     "uuid": "^3.1.0"
   },
   "resolutions": {
-    "babel-core": "^7.0.0-bridge.0"
+    "babel-core": "^7.0.0-bridge.0",
+    "@types/graphql": "^0.13.1"
```
Comment (alloy) [inline package.json]: For some weird reason a few of our deps pull in this types package as a runtime dependency, which leads to there being multiple versions… It’s really a bug upstream, but whatevs, can’t fight every battle 🤷‍♂️

File: src/lib/__tests__/graphqlTimeoutMiddleware.test.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    directive @timeout(
+      ms: Int!
+    ) on FIELD_DEFINITION
+
+    schema {
+      query: Query
+    }
+
+    type Query {
+      artwork: Artwork @timeout(ms: ${artworkTimeout})
```
Comment (alloy) [inline src/lib/__tests__/graphqlTimeoutMiddleware.test.ts]: Here’s an example of explicit timeout for a resolver.

**Comment by alloy** [comment]
Note that this will not cancel side-effects of in process resolvers, it will only stop the resolver and return `null` meaning that nested queries will no longer be executed.

File: src/lib/graphqlTimeoutMiddleware.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import { IMiddleware } from "graphql-middleware"
```
Comment (mzikherman) [inline src/lib/graphqlTimeoutMiddleware.ts]: 👋 hi middleware. IMatt

**Comment by mzikherman** [comment]
This looks great to me, nice tests! Interesting use of `directive` to enable the override on a per field basis as well.

Let's hold this prior to merging since we want to test and deploy the new Datadog integration, currently on staging.

File: src/lib/graphqlTimeoutMiddleware.ts (hunk 3 above)
Comment (alloy) [reply src/lib/graphqlTimeoutMiddleware.ts]: Yeah this is a very ugly (IMO) TS convention that indicates this is an interface. 🤷‍♂️

**Comment by alloy** [comment]
Speaking of DataDog integration, is there some way to surface in DD that a query has timed out? Otherwise it may skew aggregates incorrectly? /cc @saolsen

**Comment by alloy** [comment]
Here we go!

**Pull request merged**